core/src/mast/debuginfo/mod.rs

```diff
@@ -1,7 +1,258 @@
+    }
+
+    /// Creates an empty DebugInfo with specified capacities.
+    pub fn with_capacity(
+        decorators_capacity: usize,
+        nodes_capacity: usize,
+        operations_capacity: usize,
+        decorator_ids_capacity: usize,
+    ) -> Self {
+        Self {
+            decorators: IndexVec::with_capacity(decorators_capacity),
+            op_decorator_storage: OpToDecoratorIds::with_capacity(
+                nodes_capacity,
+                operations_capacity,
+                decorator_ids_capacity,
+            ),
+            node_decorator_storage: NodeToDecoratorIds::with_capacity(nodes_capacity, 0, 0),
+            error_codes: BTreeMap::new(),
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns true if this DebugInfo has no decorators or error codes.
+    pub fn is_empty(&self) -> bool {
+        self.decorators.is_empty() && self.error_codes.is_empty()
+    }
+
+    /// Strips all debug information, removing decorators and error codes.
+    /// This is used for release builds where debug info is not needed.
+    pub fn strip(&mut self) {
+        self.clear_decorators();
+        self.decorators = IndexVec::new();
+        self.error_codes.clear();
+    }
+
+    // DECORATOR ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of decorators.
+    pub fn num_decorators(&self) -> usize {
+        self.decorators.len()
+    }
+
+    /// Returns all decorators as a slice.
+    pub fn decorators(&self) -> &[Decorator] {
+        self.decorators.as_slice()
+    }
+
+    /// Returns the decorator with the given ID, if it exists.
+    pub fn decorator(&self, decorator_id: DecoratorId) -> Option<&Decorator> {
+        self.decorators.get(decorator_id)
+    }
+
+    /// Returns the before-enter decorators for the given node.
+    pub fn before_enter_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_before_decorators(node_id)
+    }
+
+    /// Returns the after-exit decorators for the given node.
+    pub fn after_exit_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_after_decorators(node_id)
+    }
+
+    /// Returns decorators for a specific operation within a node.
+    pub fn decorators_for_operation(
+        &self,
+        node_id: MastNodeId,
+        local_op_idx: usize,
+    ) -> &[DecoratorId] {
+        self.op_decorator_storage
+            .decorator_ids_for_operation(node_id, local_op_idx)
+            .unwrap_or(&[])
+    }
+
+    /// Returns decorator links for a node, including operation indices.
+    pub fn decorator_links_for_node(
+        &self,
+        node_id: MastNodeId,
+    ) -> Result<DecoratedLinks<'_>, DecoratorIndexError> {
+        self.op_decorator_storage.decorator_links_for_node(node_id)
+    }
```

Comment:
Not for this PR, but eventually, I'd love to get rid of this method.

Reply:
What's the rationale? Incidentally this is a pure getter.

Reply:
I meant this for a longer-term goal as we get rid of the "own" vs. "linked" decorator concepts. Though, maybe even sooner than that we can get rid of the `BasicBlock::indexed_decorator_iter()` which I think is the primary user of this method.